replace stationary by ONONSPEC projection

robinhasse opened the pull request:
This PR introduces a new projection of non-specified IEA balance flows that are currently mapped to buildings. A different allocation across sectors proposed in #609 is not subject of this PR.

# status quo

There is a very old file that has SSP1 - SSP5 scenarios for all demand mapped to industry and buildings. All energy demand that is mapped to buildings but not represented in EDGE-B currently comes from this file as "stationary". In `convertStationary`, the outdated historic demand from the file is replaced by current IEA balance data that linearly transitions to the projections in the file when going into the future. The respective flows belong to the IEA flow `ONONSPEC` - other non-specified demand. Even though we have no idea what this is, there are different projections either approaching zero or higher values. In the most striking example, we see exponentially increasing historic electricity demand in CHN but a decay towards zero in the projections.

# change

This PR gets rid of the file overall. Instead we now project the IEA demand dynamically based on the most recent historic observations. As we have no idea, where the demand comes from, I want to avoid to make too specific assumptions for the future, so we just continue with the recent slope and then approach a constant value.

## method

We assume an asymptotic model. It start from the level $x_0$ at the end of history (EOH) and with the EOH slope $\dot{x}_0$ and approaches the fraction $\epsilon$ of this slope within $\Delta t$. Both $x_0$ and $\dot{x}_0$ are determined through a linear regression of the last $n$ time steps.

$x(t) = x_0 + \dfrac{\dot{x}_0}{c} \cdot [ 1 - \exp ( - c \cdot ( t - t_0 ) ) ]$
with the decay rate $c = -\dfrac{\ln \epsilon}{\Delta t}$

The method requires the following parameters. I indicated my current choice
- $n = 5$: not as relevant, I think. 
- $\epsilon = 0.05$:  tolerance to considere the slope constant. The higher this parameter, the higher (lower) are approached long-term values for positive (negative) EOH slope.
- $\Delta t = 15 - 25 yr$: time span within which slope has decreased to a value considered constant: $\dot{x}(t_0 + \Delta t) = \epsilon \cdot \dot{x}_0$

## scenarios

I had a hard time thinking of a scenario differentiation for the projections as we had them with stationary before. @robertpietzcker, do you think this is necessary? As a first example, I created a `low`, `med` and `high` scenario referring to the long-term projection level by differentiating $\Delta t$. We can adjust the parameterisation and would have to map the level scenarios to SSPs.

| scenario | $\Delta t$, if $\dot{x}_0 > 0$ | $\Delta t$, if $\dot{x}_0 \le 0$ |
| ---- | ---- | ---- |
| low | 15yr | 25yr |
| med | 20yr | 20yr |
| high | 25yr | 15yr |

This is a comparison of the projections showing the new scenarios as thick lines and the old ones in thin lines. Mind that the old scenarios are without my hot fix in #635.
![ononspecComparison](https://github.com/user-attachments/assets/e5773acf-546a-4c83-8a48-76184e2d39f2)

## implementation

This one is for @fbenke-pik: I currently get the relevant IEA flows through `calcOutput("IOEdgeBuildings", subtype = "output_EDGE")`. This gives me the stationary IEA flows in buildings. From these, I still filter only those flows that are not represented in EDGE-B. This works fine but seems a bit intransparent. It might be better to update the mapping `structuremappingIO_outputs.csv` to get just the flows we need here and not the ones represented in EDGE-B. Not sure if this would come with an overall change to `mrcommons:calcIOEdgeBuildings` and what cache implications would be.

ricardarosemann commented on R/calcFeDemandONONSPEC.R:
```diff
@@ -0,0 +1,112 @@
+    periodEOH <- max(x$period)
+    coefs <- data.frame(periodEOH = periodEOH,
+                        slopeEOH = coef(m)[["period"]],
+                        valueEOH = predict(m, list(period = periodEOH)))
```
```suggestion
                        valueEOH = predict(m, data.frame(period = periodEOH)))
```
Not sure if this makes much of a difference, but usually ```predict``` takes data frames.

robinhasse replied:
It works with both and I used a list here as I only pass one number.

ricardarosemann reviewed (commented):
Thanks, the implementation looks good and makes sense and I find most of the resulting projections reasonable (with few exceptions). Let me know if you want me to look at some more details.

robinhasse commented:
This is for $n=10$. Not radically different but outliers like in `REF` `feothliquid` distort less.
![ononspecComparison_n10](https://github.com/user-attachments/assets/7ebc9b50-f7cf-4804-b1cd-ceac533fafb5)

robertpietzcker commented:
yes, the second version looks better, especially when looking at REF :-) 
I like it

robinhasse commented:
Thanks @robertpietzcker. If you could quickly answer these questions (or approve my suggestions). Thanks!
- Do you think, it is worth keeping a scenario differentiation in the projections?
- Do the 3 scenarios differentiated by long-term level make sense to you and are you happy with the parameterisation?
- How would you map the level scenarios to the SSPs? This would roughly fit total FE demand in EDGE-B scenarios:
| SSP scenario | level scenarios |
| ---- | ---- |
| SSP1 | low |
| SSP2 | med |
| SSP3 | med |
| SSP4 | med |
| SSP5 | high |

robertpietzcker commented:
yes, looks good to me.

robertpietzcker reviewed (approved):
I can't comment on the R code, but the results look good

robinhasse commented:
# additional changes

## scenario mapping

In the recent version, I used $n = 10$ and mapped the scenarios as suggested by Robert:

| SSP scenario | level scenario |
| ---- | ---- |
| SSP1 | low |
| SSP2_lowEn | low |
| SSP3 | high |
| SSP5 | high |
| SSP2IndiaHigh| high |
| *all others* | med |

## identical end of history

I adapted the projection such that als scenarios are identical until the end of history (EOH) in 2025. Between end of data (EOD) and EOH, we use the `med` parameterisation. Afterwards we extrapolate from there such that the respective scenario still reaches a slope considered constant until $\Delta t$ after EOD. The method is extensively documented in `calcFeDemandONONSPEC`. I hope it understandable. Compared to the last plot, changes in the projections are negligible. 

## cross-sectoral FE demand

I removed stationary overall and replaced it with the ONONSPEC projection. As a consequence, the FE demand from buildings lost the item `feh2s`. I dropped the stationary items `feh2s` and `fehes` for both industry and buildings as they haven't been read into REMIND for a long time and the values can't be trusted anyways. E.g., we had values greater zero for `feh2s` from buildings even though we have zero H2 in buildings in NPi scenarios.

# validation

I created and compared the following objects:
- `f_fedemand.cs4r`
- `f_fedemand_build.cs4r`
- `f36_uedemand_build.cs4r`
They differ of course in the expected buildings items: `fehob`, `esob`, `fegab`, `feheb`, `uecwb` (should be removed from mapping), `feelcb` and `feelb`. All other items, esp. the indutry items, are identical.

johanneskoch94 commented on R/calcFEdemand.R:
```diff
@@ -1,24 +1,24 @@
 #' Calculates Final Energy Demand for Industry and Buildings
 #' @param scenario GDP and pop scenarios. Passed to [mrdrivers::calcGDP()].
 #' @author Falk Benke
+#' @importFrom madrat calcOutput
```
This line https://github.com/pik-piam/mrremind/blob/a6fbedce6347f4c3d414f38f9def97aa5059072b/R/imports.R#L3 takes care off all imports from madrat and magclass, so it is just unnecessary... but that's just how I like to do things! :)

robinhasse commented:
@fbenke-pik and coordinated to not remove stationary in this PR as it is nominally used in industry still. Bu we believe it only feeds into the stationary items I remove here, so it should be safe to drop it. But this will be subject of another PR.

robinhasse commented on R/calcFeDemandONONSPEC.R:
```diff
@@ -0,0 +1,180 @@
+#'
+#' Project the IEA flow ONONSPEC into the future. As we have no idea, where this
+#' energy demand comes from, we use a very generic methos to project it into the
+#' future: We assume an asymptotic model. #' It starts from the level
```
Well spotted - thanks.

robinhasse commented on R/calcFeDemandONONSPEC.R:
```diff
@@ -0,0 +1,180 @@
+  }
+
+
+  .calcAsymptotic <- function(xStart, .xStart, dt, t, tStart, epsCorrection = 0,
```
Makes sense. I changed it to `xDotStart`.

robinhasse commented on R/calcFeDemandONONSPEC.R:
```diff
@@ -0,0 +1,177 @@
+                                           t = .data$period,
+                                           tStart = eoh,
+                                           epsCorrection = -(eoh - .data$eod) / yearsUntilConst[2]),
+             value = ifelse(.data$period > eoh, .data$valueFuture, .data$valueFuture)) %>%
```
Oh wow, I'm surprised this slipped through. Thanks!

robinhasse commented on R/calcFeDemandONONSPEC.R:
```diff
@@ -0,0 +1,177 @@
+#' with the decay rate
+#' \eqn{c_s = -\dfrac{\ln \varepsilon }
+#'                   {\Delta t_s - (t_\text{EOH} - t_\text{EOD})}
+#'      \cdot \left(1 + \dfrac{t_\text{EOH} - t_\text{EOD}}
+#'                            {\Delta t_\text{med}} \right)}
```
Thanks

ricardarosemann reviewed (approved):
All good now, thanks so much!
